vars/mailSendNotification.groovy

```diff
@@ -134,42 +134,47 @@ def getCulpritCommitters(config, currentBuild) {
 }
 
 def getCulprits(config, branch, numberOfCommits) {
-    if (branch?.startsWith('PR-')) {
-        //special GitHub Pull Request handling
-        deleteDir()
-        sshagent(
-            credentials: [config.gitSshKeyCredentialsId],
-            ignoreMissing: true
-        ) {
-            def pullRequestID = branch.replaceAll('PR-', '')
-            def localBranchName = "pr" + pullRequestID;
-            sh """git init
-git fetch ${config.gitUrl} pull/${pullRequestID}/head:${localBranchName} > /dev/null 2>&1
-git checkout -f ${localBranchName} > /dev/null 2>&1
-"""
-        }
-    } else {
-        //standard git/GitHub handling
-        if (config.gitCommitId) {
+    try {
+        if (branch?.startsWith('PR-')) {
+            //special GitHub Pull Request handling
             deleteDir()
             sshagent(
                 credentials: [config.gitSshKeyCredentialsId],
                 ignoreMissing: true
             ) {
-                sh """git clone ${config.gitUrl} .
-git checkout ${config.gitCommitId} > /dev/null 2>&1"""
+                def pullRequestID = branch.replaceAll('PR-', '')
+                def localBranchName = "pr" + pullRequestID;
+                sh """git init
+    git fetch ${config.gitUrl} pull/${pullRequestID}/head:${localBranchName} > /dev/null 2>&1
+    git checkout -f ${localBranchName} > /dev/null 2>&1
+    """
             }
         } else {
-            def retCode = sh(returnStatus: true, script: 'git log > /dev/null 2>&1')
-            if (retCode != 0) {
-                echo "[${STEP_NAME}] No git context available to retrieve culprits"
-                return ''
+            //standard git/GitHub handling
+            if (config.gitCommitId) {
+                deleteDir()
+                sshagent(
+                    credentials: [config.gitSshKeyCredentialsId],
+                    ignoreMissing: true
+                ) {
+                    sh """git clone ${config.gitUrl} .
+    git checkout ${config.gitCommitId} > /dev/null 2>&1"""
+                }
+            } else {
+                def retCode = sh(returnStatus: true, script: 'git log > /dev/null 2>&1')
+                if (retCode != 0) {
+                    echo "[${STEP_NAME}] No git context available to retrieve culprits"
+                    return ''
+                }
             }
         }
-    }
 
-    def recipients = sh(returnStdout: true, script: "git log -${numberOfCommits} --pretty=format:'%ae %ce'")
-    return getDistinctRecipients(recipients)
+        def recipients = sh(returnStdout: true, script: "git log -${numberOfCommits} --pretty=format:'%ae %ce'")
+        return getDistinctRecipients(recipients)
+    } catch(err) {
+        echo "[${STEP_NAME}] Culprit retrieval from git failed with '${err.getMessage()}'. Please make sure to configure gitSshKeyCredentialsId. So far, only fixed list of recipients is used."
```

Comment:
In the log message we ask for configuring ssh. But there happens much more in try block. We cannot be sure that the failure is related to ssh. Would it be possible to make the try body smaller. Would it be possible to catch more specific exceptions?

Reply:
The critical pieces are the git calls. 
Those appear in all branches of the if statement.

I had it smaller and real-life tests revealed that it was too narrow.